fx quant: do not observe bias on F.linear

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49740 fx quant: split linear test cases
* #49717 fx quant: hook up ConvTranspose{n}d
* #49688 fx quant: improve types on convert
* **#49628 fx quant: do not observe bias on F.linear**
* #49623 fx quant: do not observe bias on F.conv

Summary:

Ensures that linear bias is not observed in a `F.linear` call. This should
be a small speedup in PTQ, and will change numerics (in a good way) for
QAT if someone is using `F.linear`.

Note: the implementation is slightly more verbose compared to conv
because bias is a keyword argument in Linear.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_linear_functional_bias_not_observed
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25653532](https://our.internmc.facebook.com/intern/diff/D25653532)

vkuzo committed:
fx quant: do not observe bias on F.linear

Summary:

Ensures that linear bias is not observed in a `F.linear` call. This should
be a small speedup in PTQ, and will change numerics (in a good way) for
QAT if someone is using `F.linear`.

Note: the implementation is slightly more verbose compared to conv
because bias is a keyword argument in Linear.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_linear_functional_bias_not_observed
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
nit on "fx quant: do not observe bias on F.linear"


Summary:

Ensures that linear bias is not observed in a `F.linear` call. This should
be a small speedup in PTQ, and will change numerics (in a good way) for
QAT if someone is using `F.linear`.

Note: the implementation is slightly more verbose compared to conv
because bias is a keyword argument in Linear.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_linear_functional_bias_not_observed
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25653532](https://our.internmc.facebook.com/intern/diff/D25653532)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1347,6 +1347,20 @@ def test_conv_functional_bias_not_observed(self):
                 prepare_expected_node_occurrence=expected_node_occurrence,
             )
 
+    def test_linear_functional_bias_not_observed(self):
```
can we enable the functional test in test_linear? it was not enabled because of some issues in fx which was fixed now I think

vkuzo replied:
sure, did this in https://github.com/pytorch/pytorch/pull/49740 to keep PRs small

vkuzo committed:
Update on "fx quant: do not observe bias on F.linear"


Summary:

Ensures that linear bias is not observed in a `F.linear` call. This should
be a small speedup in PTQ, and will change numerics (in a good way) for
QAT if someone is using `F.linear`.

Note: the implementation is slightly more verbose compared to conv
because bias is a keyword argument in Linear.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_linear_functional_bias_not_observed
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25653532](https://our.internmc.facebook.com/intern/diff/D25653532)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/quantize.py:
```diff
@@ -249,21 +249,23 @@ def node_arg_is_weight(node: Node, arg: Any) -> bool:
+                    return True
+        elif node.target is torch.nn.functional.linear:
+            for kwarg_name, kwarg_value in node.kwargs.items():
+                if kwarg_name == 'bias' and arg is kwarg_value:
```
hmmm, can 'weight' be a kwarg as well?

vkuzo replied:
yeah, `weight` is a keyword argument for Linear: https://pytorch.org/docs/stable/nn.functional.html#torch.nn.functional.linear

ngimel replied:
Bias isn't kwarg-only argument (e.g. the following code is valid `torch.nn.functional.linear(x,weight, bias) `) - will this situation still be correctly handled?

vkuzo replied:
Yeah, this works with the current code.  The test case added in this PR traces through an instance of `nn.Linear`, which contains the call to `F.linear(input, self.weight, self.bias)`, with bias specified as a positional argument.

I believe the reason that it works is that when FX traces through the calls, the arguments which are defined in the function def as keyword are captured in `node.kwargs`, even if they are specified by the user as positional.  I haven't directly verified this yet.

vkuzo committed:
Update on "fx quant: do not observe bias on F.linear"


Summary:

Ensures that linear bias is not observed in a `F.linear` call. This should
be a small speedup in PTQ, and will change numerics (in a good way) for
QAT if someone is using `F.linear`.

Note: the implementation is slightly more verbose compared to conv
because bias is a keyword argument in Linear.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_linear_functional_bias_not_observed
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25653532](https://our.internmc.facebook.com/intern/diff/D25653532)

[ghstack-poisoned]
